diff --git a/lib/mdns/src/NetworkInterfaces.c b/lib/mdns/src/NetworkInterfaces.c
index 9134bd7..95e7481 100644
--- a/lib/mdns/src/NetworkInterfaces.c
+++ b/lib/mdns/src/NetworkInterfaces.c
@@ -109,11 +109,7 @@ CreateCaptureSocket(const char *ifName)
   struct sockaddr_ll bindTo;
   int skfd = 0;
   /* Open cooked IP packet socket */
-  if (olsr_cnf->ip_version == AF_INET) {
-    skfd = socket(PF_PACKET, SOCK_DGRAM, htons(ETH_P_IP));
-  } else {
-    skfd = socket(PF_PACKET, SOCK_DGRAM, htons(ETH_P_IPV6));
-  }
+  skfd = socket(PF_PACKET, SOCK_DGRAM, htons(ETH_P_ALL));
   if (skfd < 0) {
     BmfPError("socket(PF_PACKET) error");
     return -1;
@@ -142,11 +138,7 @@ CreateCaptureSocket(const char *ifName)
   /* Bind the socket to the specified interface */
   memset(&bindTo, 0, sizeof(bindTo));
   bindTo.sll_family = AF_PACKET;
-  if (olsr_cnf->ip_version == AF_INET) {
-    bindTo.sll_protocol = htons(ETH_P_IP);
-  } else {
-    bindTo.sll_protocol = htons(ETH_P_IPV6);
-  }
+  bindTo.sll_protocol = htons(ETH_P_ALL);
   bindTo.sll_ifindex = ifIndex;
   memcpy(bindTo.sll_addr, req.ifr_hwaddr.sa_data, IFHWADDRLEN);
   bindTo.sll_halen = IFHWADDRLEN;
