diff --git a/Makefile b/Makefile
index 8f39d58..2cf052a 100644
--- a/Makefile
+++ b/Makefile
@@ -235,6 +235,11 @@ watchdog:
 		$(MAKECMD) -C lib/watchdog
 		$(MAKECMD) -C lib/watchdog DESTDIR=$(DESTDIR) install
 
+dyn_hna:
+		$(MAKECMD) -C lib/dyn_hna clean
+		$(MAKECMD) -C lib/dyn_hna
+		$(MAKECMD) -C lib/dyn_hna DESTDIR=$(DESTDIR) install
+
 build_all:	all switch libs
 install_all:	install install_libs
 clean_all:	uberclean clean_libs
diff --git a/lib/dyn_hna/Makefile b/lib/dyn_hna/Makefile
new file mode 100644
index 0000000..6d0f3dc
--- /dev/null
+++ b/lib/dyn_hna/Makefile
@@ -0,0 +1,34 @@
+#
+#Copyright (C) 2010  Gioacchino Mazzurco <gmazzurco89@gmail.com>
+#
+#This program is free software: you can redistribute it and/or modify
+#it under the terms of the GNU General Public License as published by
+#the Free Software Foundation, either version 3 of the License, or
+#(at your option) any later version.
+#
+#This program is distributed in the hope that it will be useful,
+#but WITHOUT ANY WARRANTY; without even the implied warranty of
+#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#GNU General Public License for more details.
+#
+#You should have received a copy of the GNU General Public License
+#along with this file.  If not, see <http://www.gnu.org/licenses/>.
+
+OLSRD_PLUGIN =	true
+PLUGIN_NAME =	olsrd_dyn_hna
+PLUGIN_VER =	0.1
+
+TOPDIR = ../..
+include $(TOPDIR)/Makefile.inc
+
+default_target: $(PLUGIN_FULLNAME)
+
+$(PLUGIN_FULLNAME): $(OBJS) version-script.txt
+		$(CC) $(LDFLAGS) -o $(PLUGIN_FULLNAME) $(OBJS) $(LIBS)
+
+install:	$(PLUGIN_FULLNAME)
+		$(STRIP) $(PLUGIN_FULLNAME)
+		$(INSTALL_LIB)
+
+clean:
+		rm -f $(OBJS) $(SRCS:%.c=%.d) $(PLUGIN_FULLNAME)
diff --git a/lib/dyn_hna/README_dyn_hna b/lib/dyn_hna/README_dyn_hna
new file mode 100644
index 0000000..336f4dc
--- /dev/null
+++ b/lib/dyn_hna/README_dyn_hna
@@ -0,0 +1,29 @@
+---------------------------------------------------------------------
+DYNAMIC HNA PLUGIN FOR OLSRD
+by Gioacchino Mazzurco <gmazzurco89@gmail.com>
+---------------------------------------------------------------------
+
+This plugin permit to load/unload new dynamic hna on already running olsrd daemon withoiut restarting it symply writing on a fifo
+
+---------------------------------------------------------------------
+PLUGIN PARAMETERS (PlParam)
+---------------------------------------------------------------------
+
+PlParam "fifoFile" "/path/toFifoFile"
+	The location on filesystem of used fifo.
+
+
+---------------------------------------------------------------------
+SAMPLE CONFIG
+---------------------------------------------------------------------
+
+add in /etc/olsrd.conf:
+
+LoadPlugin "olsrd_dyn_hna.so.0.1"
+{
+	PlParam "fifoFile" "/var/dyn_hna.fifo"
+}
+
+
+---------------------------------------------------------------------
+EOF / 28.06.2010
diff --git a/lib/dyn_hna/src/dyn_hna.c b/lib/dyn_hna/src/dyn_hna.c
new file mode 100644
index 0000000..b1e59d3
--- /dev/null
+++ b/lib/dyn_hna/src/dyn_hna.c
@@ -0,0 +1,59 @@
+/*
+ *
+ * Copyright (C) 2010  Gioacchino Mazzurco <gmazzurco89@gmail.com>
+ * Copyright (C) 2010  Salvatore Mazzurco <s4lv00@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this file.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/* System includes */
+#include <assert.h>            /* assert() */
+#include <stddef.h>            /* NULL */
+#include <stdio.h>
+#include <sys/types.h>
+#include <stdlib.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <ctype.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+
+/* OLSRD includes */
+#include "plugin.h"
+#include "plugin_util.h"
+#include "defs.h"              /* uint8_t, olsr_cnf */
+#include "scheduler.h"          /* olsr_start_timer() */
+#include "olsr_cfg.h"          /* olsr_cnf() */
+#include "olsr_cookie.h"        /* olsr_alloc_cookie() */
+#define PLUGIN_INTERFACE_VERSION 5
+
+
+int olsrd_plugin_interface_version(void)
+{
+  return PLUGIN_INTERFACE_VERSION;
+}
+
+
+void olsrd_get_plugin_parameters(const struct olsrd_plugin_parameters **params __attribute__ ((unused)), int *size __attribute__ ((unused)))
+{}
+
+
+int olsrd_plugin_init(void)
+{
+  //custom init code
+  return 1;
+}
diff --git a/lib/dyn_hna/src/dyn_hna.h b/lib/dyn_hna/src/dyn_hna.h
new file mode 100644
index 0000000..74f77fd
--- /dev/null
+++ b/lib/dyn_hna/src/dyn_hna.h
@@ -0,0 +1,33 @@
+/*
+ *
+ * Copyright (C) 2010  Gioacchino Mazzurco <gmazzurco89@gmail.com>
+ * Copyright (C) 2010  Salvatore Mazzurco <s4lv00@gmail.com>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this file.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+ * Dynamic linked library for the olsr.org olsr daemon
+ */
+
+#ifndef _OLSRD_PLUGIN_DYN_HNA
+#define _OLSRD_PLUGIN_DYN_HNA
+
+/* Initialization function */
+int olsrd_plugin_init(void);
+
+int olsrd_plugin_interface_version(void);
+
+#endif
diff --git a/lib/dyn_hna/version-script.txt b/lib/dyn_hna/version-script.txt
new file mode 100644
index 0000000..a145659
--- /dev/null
+++ b/lib/dyn_hna/version-script.txt
@@ -0,0 +1,10 @@
+VERS_1.0
+{
+  global:
+    olsrd_plugin_interface_version;
+    olsrd_plugin_init;
+    olsrd_get_plugin_parameters;
+
+  local:
+    *;
+};
